[Bugfix][Server] WFS GetFeature GML: CDATA was url encoded before insert in GML

Gustry opened the pull request:
The CDATA section was not well formed in the GML build by QGIS Server. We found `&lt;![CDATA[ ]]&gt;`, instead of `<![CDATA[ ]]>`.

For exemple we found
```xml
<qgs:name>&lt;![CDATA[FOO &amp; FOO]]&gt;</qgs:name>
```
instead of
```xml
<qgs:name><![CDATA[FOO &amp; FOO]]></qgs:name>
```

* Funded by Faunalia

CC @pcav

pblottiere commented on src/server/services/wfs/qgswfsgetfeature.cpp:
```diff
@@ -69,6 +69,8 @@ namespace QgsWfs
 
     QString createFeatureGeoJSON( const QgsFeature &feature, const createFeatureParams &params, const QgsAttributeList &pkAttributes );
 
+    QDomElement createFieldElement( const QgsField &field, const int idx, const QgsAttributes &featureAttributes, QDomDocument &doc );
```
Could we use a `QVariant` in parameter instead of `QgsAttributes` + the corresponding index?

pblottiere replied:
And I'm in favor of the `const int` notation but only in the implementation part, not in the definition (it may be confusing from an API point of view).

pblottiere commented on src/server/services/wfs/qgswfsgetfeature.cpp:
```diff
@@ -1412,26 +1414,17 @@ namespace QgsWfs
-          fieldElem.setAttribute( QStringLiteral( "xsi:nil" ), QStringLiteral( "true" ) );
-        }
-        fieldElem.appendChild( fieldText );
+        QDomElement fieldElem = createFieldElement( fields.at( idx ), idx, featureAttributes, doc );
```
Can it be a `const` variable?

pblottiere commented on src/server/services/wfs/qgswfsgetfeature.cpp:
```diff
@@ -1528,22 +1521,38 @@ namespace QgsWfs
 
-        const QgsField field = fields.at( idx );
-        const QgsEditorWidgetSetup setup = field.editorWidgetSetup();
+        QDomElement fieldElem = createFieldElement( fields.at( idx ), idx, featureAttributes, doc );
```
const?

pblottiere commented on src/server/services/wfs/qgswfsgetfeature.cpp:
```diff
@@ -1528,22 +1521,38 @@ namespace QgsWfs
+    {
+      const QgsEditorWidgetSetup setup = field.editorWidgetSetup();
+      QString attributeName = field.name();
+      //int idx = featureAttributes.attributeIndexes[i];
```
leftover

pblottiere commented on src/server/services/wfs/qgswfsgetfeature.cpp:
```diff
@@ -1528,22 +1521,38 @@ namespace QgsWfs
+    QDomElement createFieldElement( const QgsField &field, const int idx, const QgsAttributes &featureAttributes, QDomDocument &doc )
+    {
+      const QgsEditorWidgetSetup setup = field.editorWidgetSetup();
+      QString attributeName = field.name();
```
const

Gustry replied:
There are 2 `replace` the line just after

pblottiere commented on src/server/services/wfs/qgswfsgetfeature.cpp:
```diff
@@ -1528,22 +1521,38 @@ namespace QgsWfs
+      }
+      else
+      {
+        QString fieldText = encodeValueToText( featureAttributes[idx], setup );
```
```suggestion
        const QString fieldText = encodeValueToText( featureAttributes[idx], setup );
```

pblottiere commented:
Thanks for this PR @Gustry :+1:.

Do you think that we could add a dedicated c++ test for the new `createFieldElement(...)` method? For now we only have these kind of tests for the WMS service but it could be a good start for WFS :). Let me know if you want me to take a look.

Gustry commented:
Thanks for the review @pblottiere.
I fixed all comments above.

About the C++ tests, creating a new test file in cpp is not my cup of tea. I just tried, I got some namespace issue, I'm not sure what to do.
Is the test about the query not enough ? Otherwise, I need to check with @dmarteau.

Gustry committed:
[Bugfix][Server] WFS GetFeature GML: CDATA was url encoded before insert
in GML

The CDATA section was not well formed in the GML build by QGIS Server. We found `&lt;![CDATA[ ]]&gt;`, instead of `<![CDATA[ ]]>`.

For exemple we found
```xml
<qgs:name>&lt;![CDATA[FOO &amp; FOO]]&gt;</qgs:name>
```
instead of
```xml
<qgs:name><![CDATA[FOO &amp; FOO]]></qgs:name>
```

* Funded by Faunalia

rldhont commented:
All tests arr green.

pblottiere commented on src/server/services/wfs/qgswfsgetfeature.cpp:
```diff
@@ -69,6 +69,8 @@ namespace QgsWfs
 
     QString createFeatureGeoJSON( const QgsFeature &feature, const createFeatureParams &params, const QgsAttributeList &pkAttributes );
 
+    QDomElement createFieldElement( const QgsField &field, const QVariant value, QDomDocument &doc );
```
QVariant should be a const reference.

rldhont commented:
@pblottiere thanks for the review. @Gustry has made your last request change.

pblottiere commented:
> @pblottiere thanks for the review. @Gustry has made your last request change.

Nice thanks :+1:. Feel free to merge.
